### PR TITLE
feat(wasm-gen): make first param of `gr_exit` configurable

### LINKED ISSUE #3175: Implement first param of `gr_exit` in wasm-gen to be an existing program or `gr_source`
### Problem to Solve

Currently it's always `0x00..00`.

### Possible Solution

Same as it's done for `gr_send*` sys-calls.

### Notes

_No response_

### PR DESCRIPTION
**Pull request by StackOverflowExcept1on**
Resolves #3175

@gear-tech/dev

### CODE REVIEW

File: utils/wasm-gen/src/generator/syscalls/invocator.rs [hunk 1]
```diff
@@ -303,24 +300,31 @@ impl<'a, 'b> SysCallsInvocator<'a, 'b> {
...
             Precise(SysCallName::ReservationSend),
+            Precise(SysCallName::SendCommit),
+            Precise(SysCallName::SendCommitWGas),
         ]
         .contains(&sys_call)
     }
 
+    fn is_exit_sys_call(&self, sys_call: InvocableSysCall) -> bool {
+        matches!(sys_call, InvocableSysCall::Loose(SysCallName::Exit))
+    }
```
Comment (mertwole) [inline utils/wasm-gen/src/generator/syscalls/invocator.rs:319]: can we introduce one function that will tell us to set/don't set the destination?

File: utils/wasm-gen/src/config/syscalls.rs [hunk 2]
```diff
@@ -41,7 +41,7 @@ impl SysCallsConfigBuilder {
         Self(SysCallsConfig {
             injection_amounts,
             params_config: SysCallsParamsConfig::default(),
-            sending_message_destination: MessageDestination::default(),
+            sys_call_destination: SysCallDestination::default(),
```
[the remark this answers is not in the document]
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/config/syscalls.rs:44]: It doesn't really fit here. If we use the `gr_send` then "message" destination make sense, but `gr_exit` doesn't send any messages

File: utils/wasm-gen/src/generator/syscalls/invocator.rs (hunk 1 above)
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/generator/syscalls/invocator.rs:312]: done in 91134244f529aee6c23c3d44eb85d3a0e2c26999

File: utils/wasm-gen/src/config/syscalls.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (SabaunT) [reply utils/wasm-gen/src/config/syscalls.rs:44]: maybe change that to `ExistingProgram`/`PublicPrograms`/`RecognizedPrograms`

[the remark this answers is not in the document]
Comment (SabaunT) [reply utils/wasm-gen/src/config/syscalls.rs:44]: That will state that the data is more general purpose

File: utils/wasm-gen/src/generator/syscalls.rs [hunk 3]
```diff
@@ -42,8 +42,19 @@ pub use additional_data::*;
...
 pub use invocator::*;
 
+use crate::generator::CallIndexesHandle;
 use gear_wasm_instrument::syscalls::{ParamType, SysCallName, SysCallSignature};
 
+/// Additional information about [`InvocableSysCall`].
+///
+/// Mainly used as an argument when building a sys-call.
+/// It is also possible to modify the signature while building sys-call.
+pub struct InvocableSysCallData {
```
Comment (SabaunT) [inline utils/wasm-gen/src/generator/syscalls.rs:52]: honestly, I don't this such "dto"s are suitable. It has only one purpose - carry data. The type has only three fields. Using that makes API more ambiguous. I'd get rid of it.

File: utils/wasm-gen/src/generator/syscalls/invocator.rs [hunk 4]
```diff
@@ -303,24 +293,35 @@ impl<'a, 'b> SysCallsInvocator<'a, 'b> {
...
+
+    fn is_sys_call_with_destination(&self, sys_call: InvocableSysCall) -> bool {
```
Comment (SabaunT) [inline utils/wasm-gen/src/generator/syscalls/invocator.rs:314]: move it over other `is_*` methods

File: utils/wasm-gen/src/generator/syscalls/invocator.rs [hunk 5]
```diff
@@ -199,51 +200,43 @@ impl<'a, 'b> SysCallsInvocator<'a, 'b> {
...
+                " -- Generating build call for {} sys-call with destination",
                 invocable.to_str()
             );
-            return self.build_call(
-                signature,
-                fallible,
-                insert_error_processing,
-                call_indexes_handle,
+
+            self.build_sys_call_with_destination(call_data)
```
Comment (SabaunT) [inline utils/wasm-gen/src/generator/syscalls/invocator.rs:218]:
```suggestion
            self.build_call_with_destination(call_data)
```

maybe like this? As the lower is just `build_call`, not `build_sys_call`.

**Review by SabaunT — APPROVED** [review]

**Pull request merged**